### PR TITLE
Apply ACL controls to PUBLISH operations

### PR DESCRIPTION
**Pull request by sjlongland**
This is a work-in-progress, but basically the idea is that the `topic_acl` plug-in should be able to also dictate what topics a user may publish to, as well as what topics they may subscribe from.

Not fully tested at this stage… still trying to figure out the structure of the test suite.

### CODE REVIEW

**Comment by FlorianLudwig** [comment]
Hi @sjlongland Thanks for taking this up!

Have you seen https://github.com/Yakifo/amqtt/pull/42?

I guess your work superseeds that MR.

File: tests/plugins/test_topic_checking.py
Comment (FlorianLudwig) [inline tests/plugins/test_topic_checking.py:1]: this file was neither created in 2015, nor by nicolas.

please remove this line :)

File: amqtt/broker.py
Comment (FlorianLudwig) [inline amqtt/broker.py:717]: I would prefer to see an `enum` defined for `action` - this way a type checker warns if an unknown action would be passed some place.

File: amqtt/plugins/topic_checking.py
Comment (FlorianLudwig) [inline amqtt/plugins/topic_checking.py:73]:
```suggestion
            if action == "publish" and "publish-acl" not in self.topic_config:
```

the parentheses are not needed

File: tests/plugins/test_topic_checking.py
Comment (sjlongland) [reply tests/plugins/test_topic_checking.py:1]: Heh, I did a copy from `test_authentication.py`… but I suppose I've thrown out enough for it to not be considered a "derivative work".  Question is, what should appear in the header?

Something like this?
```
# Copyright (c) 2021 aMQTT Contributers (https://github.com/Yakifo/amqtt/graphs/contributors)
# SPDX-License-Identifier: MIT
```

File: amqtt/broker.py
Comment (sjlongland) [reply amqtt/broker.py:717]: Fair point, I'll make that amendment… I was initially thinking of using the KISS approach, but you're right, it's really an `Enum`.

Where's a good place to define the `Enum` itself?

File: amqtt/plugins/topic_checking.py
Comment (sjlongland) [reply amqtt/plugins/topic_checking.py:73]: Heh, blame me bouncing between C, C++, JavaScript, Java and Python thus not always remembering operator precedence and basically brute-forcing my intentions… I'll clean that up. :-)

**Comment by sjlongland** [comment]
> Hi @sjlongland Thanks for taking this up!
> 
> Have you seen #42?
> 
> I guess your work superseeds that MR.

I had not seen that (should've looked first I guess).  Seems there's a bit more to be done on that before it can be merged.

**Comment by sjlongland** [comment]
I couldn't figure out an easy way to "unit test" the relevant code paths in `amqtt/broker.py`, so wound up spinning up a separately configured server on a different port.  Not ideal, but it gets the job done, and the plug-ins themselves have unit tests now.

File: tests/plugins/test_topic_checking.py
Comment (FlorianLudwig) [reply tests/plugins/test_topic_checking.py:1]: IANAL - so this is no legal advice just what I found online and my own understanding of navigating copyright laws:

There is no legal reason to have these lines at all. Therefore I would actually prefer not have them at all.  We do not have to remind the reader of the code in every file what the license is...

But out of respect (and not legal requirement) to the original owner of hbmqtt I would not go as far as to remove them all over the code base as it is really close to what he wrote.

Comment (FlorianLudwig) [reply tests/plugins/test_topic_checking.py:1]: @HerrMuellerluedenscheid your take?

File: tests/plugins/test_topic_checking.py
Comment (FlorianLudwig) [inline tests/plugins/test_topic_checking.py:32]: I like the dummy logger approach.

Do you think it would make sense to make a pytest fixture out of the context? It would provide the base context, with the DummyLogger and an empty config.  So each test just has to change the config instead of creating the context and logger.

File: tests/plugins/test_topic_checking.py
Comment (FlorianLudwig) [inline tests/plugins/test_topic_checking.py:80]: We should also have a test with `enabled: False`

File: amqtt/broker.py
Comment (FlorianLudwig) [reply amqtt/broker.py:717]: I would keep it inside the `broker.py`

File: tests/plugins/test_topic_checking.py
Comment (HerrMuellerluedenscheid) [reply tests/plugins/test_topic_checking.py:1]: Yes, agreed. Please omit these on newly added files.

File: tests/plugins/test_topic_checking.py
Comment (sjlongland) [reply tests/plugins/test_topic_checking.py:266]: Ahh okay, I've tried using `await` in an expression and had syntax errors thrown at me, so I got used to "just stow it in a variable" as a work-around.

File: tests/plugins/test_topic_checking.py
Comment (sjlongland) [reply tests/plugins/test_topic_checking.py:80]: I don't think any of the code paths actually look at the `enabled` flag, but I'll put it in anyway. :-)

File: tests/plugins/test_topic_checking.py
Comment (sjlongland) [reply tests/plugins/test_topic_checking.py:32]: That dummy logger is very minimal, to be properly useful I think I'd need to implement the full `logging.Logger` interface.  Either that, or hook a logging channel into the `logging` framework that makes the log messages available as a variable.

In fact, I think I favour the latter approach for a suite-wide dummy logger, since it'll also work for the classes where a logger was created with `logging.getLogger(__none__)`.  I'll have a look.

File: tests/plugins/test_topic_checking.py
Comment (sjlongland) [reply tests/plugins/test_topic_checking.py:65]: Heh, copy+paste betrays me again…

File: tests/plugins/test_topic_checking.py
Comment (sjlongland) [reply tests/plugins/test_topic_checking.py:28]: I did a bit of research, just to see if there was a library that implemented some sort of "stub" `logging` module.  Looking in `pypi` was a dead loss… but I did stumble on this:

http://www.simeonvisser.com/posts/how-to-mock-logging-in-python.html

I'll have to look more closely at the `mock` module in Python to see what assert statements are available, but maybe this approach might work better for use as a dummy logger fixture?

**Comment by sjlongland** [comment]
So, rather than making `DummyLogger` a fixture, I had another look around, and found a `py.test` plug-in that basically allows us to do the same thing with much greater flexibility.

https://pypi.org/project/pytest-logdog/

Have a look, and see what you think.

**Comment by FlorianLudwig** [comment]
Hi @sjlongland thanks for continuing the work on this!

The amount of tests is amazing :)

### regarding `pytest-logdog`

I usually don't test log output, so I am not really knowladable about the different log-capture plugins that exist for pytest.
This one seems to be rather newesh from a lone developer - which is something I am usually a bit cautions about.
Otherwise the plugin looks good.


### releasing
Your contribution will make it into `0.11` as it does break the API. With `0.11` we will take more freedom to break with `hbmqtt`

File: tests/plugins/passwd
Comment (FlorianLudwig) [inline tests/plugins/passwd:4]: I think we should have different passwords for different test users - otherwise a bug where `user1` can authenticate with password of `user2` might not be catched.

While this MR is not about authentication, having it in the example data is a liability in my eyes :)

File: tests/plugins/test_topic_checking.py
Comment (FlorianLudwig) [inline tests/plugins/test_topic_checking.py:32]: The example in the documentation interacts with `pile` after closing the context manager. Not sure how exactly it is intended to be used.

**Comment by sjlongland** [comment]
Yeah, the password thing was me being lazy… but you make a valid point.  I've changed those.

As for `logdog`… it's a valid criticism, I suppose the decision was between something like that which is maintained by one person and is "new", or doing something custom like the `DummyLogger` class I started (thus is newer), which we'd have to maintain ourselves.  `py.test` also has some logging capture fixtures, but they basically give you _everything_ and you have to filter it yourself, which was the problem that inspired `logdog`.

A release in `0.11` makes sense.  Shake off the shackles of `hbmqtt` in `0.10`. :-)

**Review by HerrMuellerluedenscheid — COMMENTED** [review]
LGTM! Thanks!

@FlorianLudwig your call. Approved from my side.

EDIT: Well, that was a little too early I guess :) Some failures need to be fixed on the pipeline.

**Comment by sjlongland** [comment]
I think I've got the formatting issues sorted, just working through how to call `flake8` to get a meaningful report.

**Comment by sjlongland** [comment]
There we go, I've ignored the warnings from the `hbmqtt` tree as I didn't touch that… anything in `tests/` and `amqtt` should be clean now. :-)

**Comment by HerrMuellerluedenscheid** [comment]
Is this good to go? LGTM. @sjlongland if this is ready, can you request a review from @FlorianLudwig ?

**Comment by sjlongland** [comment]
> Is this good to go? LGTM. @sjlongland if this is ready, can you request a review from @FlorianLudwig ?

Yep, it is… and I have. :-)

**Review by FlorianLudwig — APPROVED** [review]

**Comment by FlorianLudwig** [comment]
The master-branch is now open for `0.11` - time to break some api! :)

**Comment by sjlongland** [comment]
Many thanks. :-)

**Pull request merged**
